Sources/System/FilePathComponents.swift

```diff
@@ -0,0 +1,345 @@
+  private func parseNextComponentStart(
+    afterComponentEnd i: Index.Storage
+  ) -> Index.Storage {
+    assert(i != endIndex._storage)
+    if !isSeparator(path.bytes[i]) {
+      assert(i == path.bytes.index(after: path.bytes.startIndex))
+      // TODO: what about when we're done parsing and we have null terminator?
+    }
+    return path.bytes[i...].firstIndex(where: { !isSeparator($0) }) ?? endIndex._storage
+  }
+
+  public func index(after i: Index) -> Index {
+    let end = parseComponentEnd(startingAt: i._storage)
+    if Index(end) == endIndex {
+      return endIndex
+    }
+    return Index(parseNextComponentStart(afterComponentEnd: end))
+  }
+
+  // Find the start of the component prior to the  after the end of the prior at `i`
+  private func parseComponentStart(
+    endingAt i: Index.Storage
+  ) -> Index.Storage {
+    assert(i != startIndex._storage)
+
+    return path.bytes[i...].firstIndex(where: { !isSeparator($0) }) ?? startIndex._storage
+  }
+
+  // Chew through separators until we get to a component end
+  private func parseComponentEnd(fromStart i: Index) -> Index.Storage {
+    let slice = path.bytes[..<i._storage]
+    return slice.lastIndex(where: { isSeparator($0) }) ?? startIndex._storage
+  }
+
+  public func index(before i: Index) -> Index {
+    var slice = path.bytes[..<i._storage]
+    while let c = slice.last, isSeparator(c) {
+      slice.removeLast()
+    }
+    while let c = slice.last, !isSeparator(c) {
+      slice.removeLast()
+    }
+
+    return Index(slice.endIndex)
+  }
+
+  public subscript(position: Index) -> FilePath.Component {
+    let i = position
+    let end = parseComponentEnd(startingAt: i._storage)
+    return FilePath.Component(path.bytes[i._storage ..< end])
+  }
+}
+
+// @available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)
+extension FilePath.ComponentView: RangeReplaceableCollection {
+  public init() {
+    self.init(path: FilePath())
+  }
+
+  public mutating func replaceSubrange<C: Collection>(
+    _ subrange: Range<Index>, with newElements: C
+  ) where Element == C.Element {
+    let (lowerBound, upperBound) = (subrange.lowerBound, subrange.upperBound)
+
+    let pathRange = lowerBound._storage ..< upperBound._storage
+    guard !newElements.isEmpty else {
+      path.bytes.removeSubrange(pathRange)
+      return
+    }
+
+    // Insertion skips roots
+    let hasNewComponents = !newElements.lazy.filter { !$0.isRoot }.isEmpty
+
+    // Explicitly add a trailing separator if
+    //   not at end and next character is not a separator
+    let atEnd = upperBound == endIndex
+    let trailingSeparator = !atEnd && !isSeparator(path.bytes[upperBound._storage])
+
+    // Explicitly add a preceding separator if
+    //   replacing front with absolute components (unless redundant by trailing separator),
+    //   preceding character is not a separator (which implies at the end)
+    let atStart = lowerBound == startIndex
```

Comment:
TODO: Account for `//` special case...

> Multiple successive <slash> characters are considered to be the same as one <slash>, except for the case of exactly two leading <slash> characters.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_271

Reply:
For POSIX, it seems to denote implementation defined path resolution:

> A pathname consisting of a single <slash> shall resolve to the root directory of the process. A null pathname shall not be successfully resolved. If a pathname begins with two successive <slash> characters, the first component following the leading <slash> characters may be interpreted in an implementation-defined manner, although more than two leading <slash> characters shall be treated as a single <slash> character.

https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap04.html#tag_04_13

Reply:
Luckily, all signs indicate that `//` is special cased on neither Darwin nor Linux, so there is no need for us to do anything about it. It's okay to treat `//` as a synonym of `/` for the purposes of canonicalization and component extraction.